### ISSUE TITLE
Suggestion: make suffixes that return velocity as a vector, not a structure

### ISSUE THREAD

**Issue by pand5461** [KSP-KOS/KOS#2489]
Right now, one has to use `Velocity:orbit` and `Velocity:surface` to get the velocities in the inertial and rotating reference frames, respectively. This has two issues:
* false similarity to the `Orbitable` structure in having the `:orbit` suffix, whereas it returns an entirely different datatype
* break of symmetry between `Orbitable:position` and `Orbitable:velocity` suffixes.

The proposal is to change the kOS interface to explicit `:srfvelocity` and `:obtvelocity` suffixes that simply return vectors, and introduce `obtvelocityat(object, time)` and `srfvelocityat(object, time)` functions that also both simply return vectors.

**Comment by Dunbaratu** [comment]
I don't understand this claim:

* break of symmetry between Orbitable:position and Orbitable:velocity suffixes.

``position`` literally *is* ``orbitable:position`` and ``velocity`` literally *is* ``orbitable:velocity``.  As in ``velocity`` is a shorthand for ``ship:velocity``, and ``ship`` is a Vessel, which is an orbitable.

``orbitable:velocity`` is the abstraction that works the same for a vessel or a body, and in both cases it returns a type that holds both the surface and orbital variants.  In the surface variant the parent body's rotation is subtracted from the velocity so you get your velocity relative to the spot on the parent body under you, and in the orbit variant it isn't. 

I don't see the supposed difference you're talking about.

**Comment by pand5461** [comment]
I'll clarify: for any flight prediction, one needs an initial position and an initial velocity as vectors. The former is one suffix away from an object reference, the latter is two suffixes away. Is there a really good reason to do that vs. to have one-suffix aliases for the orbital and surface velocities, just like there are `prograde` and `srfprograde` directions?

The inconsistency is worse for `positionat()` and `velocityat()`. I have the "cannot use * on Scalar and OrbitableVelocity" more frequently than I'd like.

I understand that the structure mirrors the underlying code structure. But is such mirroring necessary?

**Comment by Dunbaratu** [comment]
>  Is there a really good reason to do that

The fact that the orbital versus the surface gives a different answer is only true of the velocity, not the position.  If I did add two different suffixes for ``ship:position:surface`` and ``ship:position:orbit``, just for "consistency", they would return the same answer as each other anyway.  That's why velocity has two types and position has one.  If you are calling that "inconsistent", well then that "inconsistency" comes from the very definition of what those things *are*.

Yes, it is possible to add shortcut suffixes that skip the middle step, for convenience.  (Not for "consistency", just for convenience of trimming things down.)

> I understand that the structure mirrors the underlying code structure. But is such mirroring necessary?

They underlying code structure is that way to mirror the physical data these things represent, not the other way around.

**Comment by pand5461** [comment]
> The fact that the orbital versus the surface gives a different answer is only true of the velocity, not the position.

I do not understand this part. Say, there is a satellite in an elliptic orbit not synchronized with the planet's rotation. It's logical to say that its orbital position is the same at each periapsis pass. What do you mean by "the orbital versus surface gives the same answer for position" in that case?

Position in a specific (rotating or non-rotating) coordinate frame can be expressed as a tuple. The velocity *in the same frame* is a tuple showing how fast the position tuple changes. Going from one reference frame to another changes *both* position and velocity tuples in a consistent way.

There is a difference between position and velocity transformations in the fact that positions only rotate whereas velocities rotate and shift. In that sense, we are indeed not interested whether the frame rotates or not when we have to transform the position, we only need the current orientation of the axes. Of course, when one changes from one inertial frame to another, both position and velocity tuples only rotate.

For velocity, what kOS outputs as "surface" and "orbital" velocities are in fact "rotated and shifted" and "rotated" tuples representing the velocity in the "prime" coordinate system. If you look carefully, you can see that `velocity:orbit` does not change as one would expect a "true" orbital velocity would, i.e. dV/dt != -mu*R/|R|^3 when the ship is below the rotating-to-inertial frame transition (the difference is quite noticeable, about 0.6 m/s^2 for a circular 100 km LKO). That's because the velocity is expressed in a new inertial coordinate system each tick.

**Comment by Dunbaratu** [comment]
None of that has anything to do with what you said in this issue.

You opened the issue claiming that splitting velocity into separate surface velocity and orbital velocity suffix Instead of having to go one more structure deep to get it would somehow fix some supposed inconsistency.

What you're talking about now is about KSP's axes moving with the planet's orbit instead of staying fixed to the universe, which doesn't have a single thing to do with whether you obtain that velocity with :velocity:orbit versus :obtvelocity.

GIVEN that you wanted :obtvelocity to give the same answer as :velocity:orbit, absolutely noneof what you just said is relevant to THAT issue.

If this new second issue is what you want, then that requires a different issue.  Because changing the meaning of what you get when you ask for orbital velocity has nothing to do with what chain of suffixes you use to get to it.  What you're actually asking for seems to be to change velocity to a solarprimevector-based set of axes, which wouldn't have happened at all if I simply did what the issue originally asked for and made ``:obtvelocity`` into an alias for the current result you get from ``:velocity:orbit``.

**Comment by Dunbaratu** [comment]
Also, making the position fixed in space so it matches the velocity being fixed in space (which would be needed to *truly* have position and velocity consistent in the same reference frame) would be a massive problem for a lot of already written scripts that presume V(0,0,0) is the ship.  (i.e. they can assume that target:position:mag is the distance to the target right now, an assumption that is totally wrong if v(0,0,0) is fixed in space rather than moving with the ship.)

The ugly fact that position uses an origin that moves along with the ship, while velocity assumes the ship is in motion, and thus they are not the same reference frame, is baked into KSP at a deep level because it's how they dealt with the fact that floating-point numbers get imprecise at larger magnitude values.  Fixing the origin point for position in space would mean putting the ship at large positions like V(1000000000,1000000000,1000000000) and at numbers like those, 32-bit floats aren't very precise.  (The game itself uses 64-bit floats for its purposes, but Unity uses 32-bit when talking to the graphics card, so at some point these numbers have to be converted down into 32-bit to be handed off to the graphics card that calculates collisions, and it gives a lot of false positives for colliding parts when the magnitudes are that big and thus imprecise.  That's where the Space Kraken came from, and it's why the position is in a reference frame that moves with the ship even though that's not consistent with the reference frame they use for velocity where the velocity vector says the ship is moving at 2380 m/s.

**Comment by pand5461** [comment]
> There are two different issues:

I'm talking about (1), as fixing (2) consistently is a huge change affecting everything about steering, flight prediction etc.

The inconsistency I'm seeing *in naming* is that (a) orbital and surface velocities are packed into a structure whilst positions in inertial and rotating frames are not discriminated at all, and srf/obt prograde are two different suffixes and (b) `Orbitable:orbit` and `OrbitableVelocity:orbit` suffixes return different datatypes.

That often causes little inconvenience when one only uses `object:velocity`. However, it gives rise to expressions like `object:orbit:velocity:orbit` which is just ugly, and it makes similarly-looking functions `positionat()` and `velocityat()` return different datatypes. I don't know how large is a portion of users who uses `velocityat()`, but in my case it's done exclusively for passing the result into an orbital solver, so that the need of additional `:orbit` suffix to the velocity feels superfluous.

On the consistency between the velocity and the rate of change in the position, I'm OK with the idiom that the "velocity" is actually the rate at which `object:position - body:position` changes. As such, just having coordinate axes that either consistently rotate or consistently stay fixed is fine to me. But that's another issue entirely.

**Comment by Dunbaratu** [comment]
I do not agree at all that -1-(a) is a naming inconsistency.  More on that later.

But about -1-(b) I agree.  The two types of direction quaternion are accessed through two different suffixes that each return a different quaternion, while the two types of velocity vector are accessed through one suffix that in turn returns a tuple of two vectors.  That is the only inconsistency I see that has anything to do with naming.  Both for Directions and Velocities you have two values, one for surface and one for orbital, but the means to access those two values is different.  But to be fair, this is NOT the "inconsistency" you were talking about when you opened this issue.  You were talking about -1-(a), a discrepancy between *position* and velocity, rather than between *direction* and velocity.  That's why I brought up issue -2-, as it's the real cause of *that* discrepancy that you were first mentioning.

The reason I don't agree with you about (a) is because the naming difference was the *effect of* the difference between how velocity and position are handled, not the *cause of* it.   The fact that there only is one kind of position offered but two kinds of velocity offered is *why* velocity was given that extra suffix to get at the specific one, and position was not.  The proposed change this issue asked for does not get rid of this discrepancy at all.  All it does it transform it into a slightly differently named version of the exact same discrepancy.  Instead of "Why does velocity have an additional suffix distinguishing orbital vs surface and position does not?" it would become "Why are there two velocity suffixes called ``srfvelocity`` and ``obtvelocity`` when there's only one suffix called ``position``,  with no ``srfposition`` or ``obtposition``?"

Thus my view that what inconsistency exists is caused entirely by issue -2-.  Issue -1-a is the *effect* of this, not the cause of it.

But now that you bring up issue -1-b, which you didn't mention at first, that *is* a legitimate reason to want the change you propose.  Because unlike with position, with directions there *are* indeed two different types the script can access, orbital versus surface, for both directions and velocities, and *there* the naming convention difference between how you access them for direction versus velocity vector *is* indeed an actual naming inconsistency that's not necessary and not caused purely by the actual underlying data itself like the difference between position and velocity is.
